app/scripts/controllers/network/test/fake-provider.ts

```diff
@@ -0,0 +1,206 @@
+ */
+// NOTE: We shouldn't need to extend from the "real" provider here, but
+// we'd need a `SafeEventEmitterProvider` _interface_ and that doesn't exist (at
+// least not yet).
+export class FakeProvider extends SafeEventEmitterProvider {
+  calledStubs: FakeProviderStub[];
+
+  #originalStubs: FakeProviderStub[];
+
+  #stubs: FakeProviderStub[];
+
+  /**
+   * Makes a new instance of the fake provider.
+   *
+   * @param options - The options.
+   * @param options.stubs - A set of objects that allow specifying the behavior
+   * of specific invocations of `sendAsync` matching a `method`.
+   */
+  constructor({ stubs = [] }: FakeProviderEngineOptions) {
+    super({ engine: new JsonRpcEngine() });
+    this.#originalStubs = stubs;
+    this.#stubs = this.#originalStubs.slice();
+    this.calledStubs = [];
+  }
+
+  send = (
+    payload: JsonRpcRequest<any>,
+    callback: (error: unknown, response?: JsonRpcResponse<any>) => void,
+  ) => {
+    return this.#handleSend(payload, callback);
+  };
+
+  sendAsync = (
+    payload: JsonRpcRequest<any>,
+    callback: (error: unknown, response?: JsonRpcResponse<any>) => void,
+  ) => {
+    return this.#handleSend(payload, callback);
+  };
+
+  #handleSend(
+    payload: JsonRpcRequest<any>,
+    callback: (error: unknown, response?: JsonRpcResponse<any>) => void,
+  ) {
+    if (Array.isArray(payload)) {
+      throw new Error("Arrays aren't supported");
+    }
+
+    const index = this.#stubs.findIndex((stub) => {
+      return (
+        stub.request.method === payload.method &&
+        (!('params' in stub.request) ||
+          isDeepStrictEqual(stub.request.params, payload.params))
+      );
+    });
+
+    if (index === -1) {
+      const matchingCalledStubs = this.calledStubs.filter((stub) => {
+        return (
+          stub.request.method === payload.method &&
+          (!('params' in stub.request) ||
+            isDeepStrictEqual(stub.request.params, payload.params))
+        );
+      });
+      let message = `Could not find any stubs matching: ${inspect(payload, {
+        depth: null,
+      })}`;
+      if (matchingCalledStubs.length > 0) {
+        message += `\n\nIt appears the following stubs were defined, but have been called already:\n\n${inspect(
+          matchingCalledStubs,
+          { depth: null },
+        )}`;
+      }
+
+      throw new Error(message);
+    } else {
+      const stub = this.#stubs[index];
+
+      if (stub.discardAfterMatching !== false) {
+        this.#stubs.splice(index, 1);
+      }
+
+      if (stub.delay) {
+        originalSetTimeout(() => {
+          this.#handleRequest(stub, callback);
+        }, stub.delay);
+      } else {
+        this.#handleRequest(stub, callback);
+      }
+
+      this.calledStubs.push({ ...stub });
+    }
+  }
+
+  async #handleRequest(
+    stub: FakeProviderStub,
+    callback: (error: unknown, response?: JsonRpcResponse<any>) => void,
+  ) {
+    if (stub.beforeCompleting) {
```

Comment:
The difference in these conditions is due to the stub type (above on line 42). The `beforeCompleting` property is guaranteed to exist (but it can be undefined), so TypeScript lets us check it directly. However the `response` and `implementation` properties might not exist on the object at all, so TypeScript objects to referencing the property directly, forcing us to use the `in` operator.